pkg/services/tree/container.go

```diff
@@ -0,0 +1,83 @@
+package tree
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	containerSDK "github.com/nspcc-dev/neofs-sdk-go/container"
+	cidSDK "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	netmapSDK "github.com/nspcc-dev/neofs-sdk-go/netmap"
+)
+
+type containerCache struct {
+	sync.Mutex
+	nm  *netmapSDK.Netmap
+	lru *simplelru.LRU
+}
+
+func (c *containerCache) init() {
+	c.lru, _ = simplelru.NewLRU(defaultContainerCacheSize, nil) // no error, size is positive
+}
+
+type containerCacheItem struct {
+	cnr   *containerSDK.Container
+	local int
+	nodes netmapSDK.Nodes
+}
+
+const defaultContainerCacheSize = 10
+
+// getContainerNodes returns nodes in the container and a position of local key in the list.
+func (s *Service) getContainerNodes(cid *cidSDK.ID) (netmapSDK.Nodes, int, error) {
+	nm, err := s.nmSource.GetNetMap(0)
+	if err != nil {
+		return nil, -1, fmt.Errorf("can't get netmap: %w", err)
+	}
+
+	cnr, err := s.cnrSource.Get(cid)
+	if err != nil {
+		return nil, -1, fmt.Errorf("can't get container: %w", err)
+	}
+
+	cidStr := cid.String()
+
+	s.containerCache.Lock()
+	if s.containerCache.nm != nm {
+		s.containerCache.lru.Purge()
+	} else if v, ok := s.containerCache.lru.Get(cidStr); ok {
+		item := v.(containerCacheItem)
+		if item.cnr == cnr {
```

Comment:
do you compare pointers here? could they ever be equal?

Reply:
Yes, in the case we have received the cached value.